src/core/mount-setup.c

```diff
@@ -397,9 +397,19 @@ int mount_setup(bool loaded_policy) {
          * nspawn and the container tools work out of the box. If
          * specific setups need other settings they can reset the
          * propagation mode to private if needed. */
-        if (detect_container() <= 0)
-                if (mount(NULL, "/", NULL, MS_REC|MS_SHARED, NULL) < 0)
-                        log_warning_errno(errno, "Failed to set up the root directory for shared mount propagation: %m");
+        if (mount(NULL, "/", NULL, MS_REC|MS_SHARED, NULL) < 0)
+                log_warning_errno(errno, "Failed to set up the root directory for shared mount propagation: %m");
+
```

Comment:
Ok. I suggest to add an environment variable `ROOT_PROPAGATION=shared` for systemd (pid1).

And systemd would do the `MS_SHARED`
- either when it's not running in a container (as it does already today),
- or when that variable is set to `shared`.

Then, rkt (or any container run-time that wants to have both hierarchical mount namespaces and "machinectl bind" working correctly together) would have to set `ROOT_PROPAGATION=shared` on systemd.

However, it is not so trivial anymore because systemd-nspawn has no generic mechanism to pass custom environment variables to pid1. Instead, it has a [fixed list of environment variables it can set](https://github.com/systemd/systemd/blob/v231/src/nspawn/nspawn.c#L2536):
```
        const char *envp[] = {
                "PATH=" DEFAULT_PATH_SPLIT_USR,
                NULL, /* container */
                NULL, /* TERM */
                NULL, /* HOME */
                NULL, /* USER */
                NULL, /* LOGNAME */
                NULL, /* container_uuid */
                NULL, /* LISTEN_FDS */
                NULL, /* LISTEN_PID */
                NULL, /* NOTIFY_SOCKET */
                NULL
        };
```

There is `SYSTEMD_NSPAWN_CONTAINER_SERVICE=` on nspawn that is passed to `container=` on pid1.

We could add `SYSTEMD_NSPAWN_ROOT_PROPAGATION` on nspawn that is passed to `ROOT_PROPAGATION` on pid1...

@poettering @tixxdz what do you think?